### PR TITLE
Validate versions for CCR snapshot on matching index version.

### LINKED ISSUE #104620: [CI] CcrRollingUpgradeIT testCannotFollowLeaderInUpgradedCluster failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/s6a2urt7xmby6/tests/:x-pack:qa:rolling-upgrade-multi-cluster:v8.12.0%23follower%23upgradedClusterTest/org.elasticsearch.upgrades.CcrRollingUpgradeIT/testCannotFollowLeaderInUpgradedCluster

**Reproduction line:**
```
./gradlew ':x-pack:qa:rolling-upgrade-multi-cluster:v8.12.0#follower#upgradedClusterTest' -Dtests.class="org.elasticsearch.upgrades.CcrRollingUpgradeIT" -Dtests.method="testCannotFollowLeaderInUpgradedCluster" -Dtests.seed=77E1D53FAF91D442 -Dtests.bwc=true -Dtests.locale=zh-CN -Dtests.timezone=Europe/Paris -Druntime.java=21
```

**Applicable branches:**
8.12

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.upgrades.CcrRollingUpgradeIT#testCannotFollowLeaderInUpgradedCluster`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testCannotFollowLeaderInUpgradedCluster'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.upgrades.CcrRollingUpgradeIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
junit.framework.AssertionFailedError: Expected exception ResponseException but no exception was thrown

  at __randomizedtesting.SeedInfo.seed([77E1D53FAF91D442:F9BE25B305B8CF59]:0)
  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2889)
  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2875)
  at org.elasticsearch.upgrades.CcrRollingUpgradeIT.testCannotFollowLeaderInUpgradedCluster(CcrRollingUpgradeIT.java:225)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by mosche**
RestoreService will validate the snapshot based on max index version. However, we do not increment index version on patch releases. 

To prevent attempting to restore an index of a future patch release, this change will reject the restore already when building the snapshot info from newer nodes matching the current index version.

Related to https://github.com/elastic/elasticsearch/issues/105295
Fixes https://github.com/elastic/elasticsearch/issues/104620

### CODE REVIEW

**Comment by mosche** [comment]
The `elasticsearch-ci/8.12.2 / bwc` failure is unrelated:
> [MixedClusterEsqlSpecIT » test {stats.EvalOverridingKey SYNC}](https://gradle-enterprise.elastic.co/s/zejgk2olzmdjg/tests/:x-pack:plugin:esql:qa:server:mixed-cluster:v8.12.2%23javaRestTest/org.elasticsearch.xpack.esql.qa.mixed.MixedClusterEsqlSpecIT/test%20%7Bstats.EvalOverridingKey%20SYNC%7D)

**Review by williamrandolph — APPROVED** [review]
We'll have to revisit this when we remove or change `Node#getVersion`, but I think this is good as a test fix.

**Comment by mosche** [comment]
> We'll have to revisit this when we remove or change Node#getVersion, but I think this is good as a test fix.

Do we need to reassess the plans for this given the requirements for CCR (validate patch version constraint if index version matches)? The only alternative I see would be reverting to storing the actual version in the index metadata.

**Comment by williamrandolph** [comment]
It seems to me like a very similar use case to persisted NodeMetadata. For stateful `Elasticsearch`, I don't see why we couldn't use `BuildVersion` or a class very much like it to wrap a `Version` identifier. Perhaps we could have a pluggable `CcrVersion` that wraps `Version` in stateful Elasticsearch, and wraps `IndexVersion` in stateless? Does that seem like it could work?

**Comment by williamrandolph** [comment]
> So maybe we should plan for `DiscoveryNode.getVersion()` to return `BuildVersion` mid/long term?

Possibly -- `ServerlessBuildVersion` is intended to be a no-op (assuming the orchestration we control never tries to roll back a node in place), and we may want different behavior for Serverless CCR. Or perhaps we'll control things on the orchestration layer so that we avoid CCR incompatibilities between clusters, though I'm not sure how that would actually work, which would let us use the `BuildVersion` no-op in Elasticsearch.

All of that seems beyond the scope of this PR, though, and we'll catch it when we try to get `Version` out of `DiscoveryNode`.

**Comment by thecoop** [comment]
This needs the changes in https://github.com/elastic/elasticsearch/pull/105657 (and dependent PRs) before this can be merged

**Comment by mosche** [comment]
@thecoop This should also work without bumping the version on a new release as it's falling back to build / release version if the index version matches. But happy to wait if you prefer.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java [hunk 1]
```diff
@@ -201,15 +204,40 @@ public void getSnapshotInfo(
...
             csDeduplicator.execute(
                 new ThreadedActionListener<>(threadPool.executor(ThreadPool.Names.SNAPSHOT_META), listener.map(response -> {
+                    Snapshot snapshot = new Snapshot(this.metadata.name(), SNAPSHOT_ID);
+
+                    // RestoreService will validate the snapshot is restorable based on the max index version.
+                    // However, we do not increment index version on patch releases.
+                    // To prevent attempting to restore an index of a future patch version, we reject the restore
+                    // already when building the snapshot info from newer nodes matching the current index version.
+                    IndexVersion maxIndexVersion = response.getNodes().getMaxDataNodeCompatibleIndexVersion();
+                    if (IndexVersion.current().id() == maxIndexVersion.id()) {
```
Comment (thecoop) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java:214]: `IndexVersion` itself is equatable here

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java [hunk 2]
```diff
@@ -201,15 +204,40 @@ public void getSnapshotInfo(
...
+                        for (var node : response.nodes()) {
+                            if (node.canContainData() && node.getMaxIndexVersion().id() == maxIndexVersion.id()) {
+                                BuildVersion remoteVersion = BuildVersion.fromVersionId(node.getVersion().id);
```
Comment (thecoop) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java:217]: This is using the discoverynode release version, which we want to remove at some point... If this needs to wait until we've worked out how to do CCR properly on serverless, can this have a comment indicating so?

**Comment by thecoop** [comment]
It's more the comment does not match current reality - either we can update the comment later, or wait until the build infra changes are merged

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java (hunk 2 above)
Comment (mosche) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java:217]: Is it really a serverless issue? In the case of serverless we're probably fine with index version only. I'd expect we would bump index version accordingly. IMHO, this is just to maintain the current version constraints for stateful.

Comment (mosche) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java:218]: I've added a comment, in any case this needs to be revisited when looking into release version of discovery node

**Comment by mosche** [comment]
> It's more the comment does not match current reality - either we can fix it and update the comment later, or wait until the build infra changes are merged

I've modified the comment to be more generic, so it's correct now and later

**Review by thecoop — APPROVED** [review]

**Comment by mosche** [comment]
@elasticmachine update branch

**Comment by mosche** [comment]
Bwc test failures are unrelated, see https://github.com/elastic/elasticsearch/issues/107403

**Comment by mosche** [comment]
@elasticmachine test this please

**Pull request merged**